### PR TITLE
[SYCL] Usm allocator refactor

### PR DESCRIPTION
**Pull request by igchor**
This commit makes USMAllocator unaware of the memory type on which it is operating. The user of USMAllocator is responsible
for setting appropriate options for each instance of USMAllocator.

This is the first step to making USMAllocator generic.  In the next step, I'd like to start integration with Unified Memory Allocation (oneMemory). USMAllocator would be used as a heap manager by UMA.

### CODE REVIEW

**Comment by smaslov-intel** [comment]
@igchor : rather than adding new stuff into L0 Plugin, could you first move the USM allocator into Unified Runtime please?

**Comment by igchor** [comment]
> @igchor : rather than adding new stuff into L0 Plugin, could you first move the USM allocator into Unified Runtime please?

Sure, done: https://github.com/intel/llvm/pull/7820
I've also moved some move utils from level_zero to UR so that we can implement ur_context, etc: https://github.com/intel/llvm/pull/7821

**Review by smaslov-intel — CHANGES_REQUESTED** [review]
Update after https://github.com/intel/llvm/pull/7820 is merged

**Comment by igchor** [comment]
> Update after #7820 is merged

Done.

**Comment by smaslov-intel** [comment]
@rdeodhar : please help review the change to USM allocator

File: sycl/plugins/unified_runtime/ur/usm_allocator.cpp [hunk 1]
```diff
@@ -542,13 +269,17 @@ class USMAllocContext::USMAllocImpl {
...
-  USMAllocImpl(std::unique_ptr<SystemMemory> SystemMemHandle)
-      : MemHandle{std::move(SystemMemHandle)} {
+  USMAllocImpl(std::unique_ptr<SystemMemory> SystemMemHandle,
+               USMAllocatorParameters params)
+      : MemHandle{std::move(SystemMemHandle)}, params(params) {
 
     // Generate buckets sized such as: 64, 96, 128, 192, ..., CutOff.
     // Powers of 2 and the value halfway between the powers of 2.
-    auto Size1 = USMSettings.MinBucketSize[MemHandle->getMemType()];
+    auto Size1 = params.MinBucketSize;
```
Comment (jandres742) [inline sycl/plugins/unified_runtime/ur/usm_allocator.cpp:282]: would this affect performance during this period of transition, if MinBucketSize may be different depending on the memory type? See here
https://github.com/intel/llvm/blob/8759d0728929c4e8aae88fddb06206aad5719d2b/sycl/plugins/unified_runtime/ur/usm_allocator.cpp#L101

File: sycl/plugins/unified_runtime/ur/usm_allocator.cpp [hunk 2]
```diff
@@ -568,12 +299,12 @@ class USMAllocContext::USMAllocImpl {
     return KnownSlabs;
   }
 
-  size_t SlabMinSize() {
-    return USMSettings.SlabMinSize[(*MemHandle).getMemType()];
-  };
+  size_t SlabMinSize() { return params.SlabMinSize; };
```
Comment (jandres742) [inline sycl/plugins/unified_runtime/ur/usm_allocator.cpp:302]: same comment about performance.

File: sycl/plugins/unified_runtime/ur/usm_allocator.cpp [hunk 3]
```diff
@@ -855,27 +585,34 @@ bool Bucket::CanPool(bool &ToPool) {
...
+      size_t NewPoolSize = PoolSize + SlabAllocSize();
+
+      if (OwnAllocCtx.getParams().limits->MaxSize < NewPoolSize) {
+        break;
+      }
+
+      if (OwnAllocCtx.getParams().limits->TotalSize.compare_exchange_strong(
+              PoolSize, NewPoolSize)) {
+        if (chunkedBucket)
+          ++chunkedSlabsInPool;
```
Comment (jandres742) [inline sycl/plugins/unified_runtime/ur/usm_allocator.cpp:599]: how is this protected now that the lock at the beginning of this function was removed?

File: sycl/plugins/unified_runtime/ur/usm_allocator.cpp [hunk 4]
```diff
@@ -885,11 +622,11 @@ size_t Bucket::Capacity() {
   if (getSize() <= ChunkCutOff())
     return 1;
   else
-    return USMSettings.Capacity[getMemType()];
+    return OwnAllocCtx.getParams().Capacity;
```
Comment (jandres742) [inline sycl/plugins/unified_runtime/ur/usm_allocator.cpp:625]: how we will handle different values for Capacity?

File: sycl/plugins/unified_runtime/ur/usm_allocator.cpp (hunk 1 above)
Comment (igchor) [reply sycl/plugins/unified_runtime/ur/usm_allocator.cpp:282]: No, in fact, my goal was to keep the original behavior intact. This assignment is just in a different place: https://github.com/intel/llvm/blob/51609155370086670b74ba99f3724d3d8d340852/sycl/plugins/unified_runtime/ur/usm_allocator_config.cpp

In the original implementation, those variables (MinBucketSize, Capacity, itp.) were stored in a single global structure and indexed using `[getMemType()]`. I have moved those variables to a separate class: `USMAllocatorParameters` and I'm now passing instance of this class to the USMAllocContext as you can see here: https://github.com/intel/llvm/blob/51609155370086670b74ba99f3724d3d8d340852/sycl/plugins/level_zero/pi_level_zero.cpp#L844

Basically, instead of having Structure of Arrays:

```
class SetLimits() {
   ...
   size_t MinBucketSize[MemType::All];
};
```

we now have Array of Structs:

```
class USMAllocatorParameters {
   ...
  size_t MinBucketSize;
} configs[MemType::All];
```

File: sycl/plugins/unified_runtime/ur/usm_allocator.cpp (hunk 2 above)
Comment (igchor) [reply sycl/plugins/unified_runtime/ur/usm_allocator.cpp:302]: Please see my previous answer

File: sycl/plugins/unified_runtime/ur/usm_allocator.cpp (hunk 4 above)
Comment (igchor) [reply sycl/plugins/unified_runtime/ur/usm_allocator.cpp:625]: Please see my previous answer

File: sycl/plugins/unified_runtime/ur/usm_allocator.cpp (hunk 3 above)
Comment (igchor) [reply sycl/plugins/unified_runtime/ur/usm_allocator.cpp:599]: My assumption was that the `PoolLock` was here only to protect `CurPoolSize` update (since this is a global variable, shared between all instances of USMAllocContext. This would match the comment: https://github.com/intel/llvm/blob/sycl/sycl/plugins/unified_runtime/ur/usm_allocator.cpp#L62 Now, instead of this lock I just rely on CAS to update the value.

As for the remaining variables (e.g. chunkedSlabsInPool) they should be protected by BucketLock which must be taken before calling `CanPool`

File: sycl/plugins/unified_runtime/ur/usm_allocator.cpp (hunk 1 above)
Comment (igchor) [reply sycl/plugins/unified_runtime/ur/usm_allocator.cpp:282]: The goal of this is to split the pool management logic from reading/setting the configuration since this might differ for other adapters.

Comment (jandres742) [reply sycl/plugins/unified_runtime/ur/usm_allocator.cpp:282]: thanks @igchor.

File: sycl/plugins/unified_runtime/ur/usm_allocator.cpp (hunk 3 above)
Comment (jandres742) [reply sycl/plugins/unified_runtime/ur/usm_allocator.cpp:599]: thanks @igchor. if that's the case, then yes, we would be ok.

**Comment by igchor** [comment]
> @rdeodhar : please help review the change to USM allocator

@rdeodhar would you please take a look?

**Comment by igchor** [comment]
@intel/llvm-gatekeepers @againull @smaslov-intel is there something pending to merge this?

**Comment by bader** [comment]
> @intel/llvm-gatekeepers @againull @smaslov-intel is there something pending to merge this?

**smaslov-intel ** requested changes. He must approve before it can be merged.

**Review by smaslov-intel — APPROVED** [review]
LGTM

**Comment by smaslov-intel** [comment]
> > @intel/llvm-gatekeepers @againull @smaslov-intel is there something pending to merge this?
> 
> @smaslov-intel requested changes. He must approve before it can be merged.

I think this is good to go in.

**Comment by bader** [comment]
Approval from @intel/llvm-reviewers-runtime team is required.

**Review by steffenlarsen — APPROVED** [review]
